pyoneer/layers/dropout_impl.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+
+
+class ConcreteDropout(tf.keras.layers.Wrapper):
+    """
+    Implementation of concrete dropout (Gal et al. 2017).
+
+    Args:
+        layer: Apply dropout to the inputs of `layer` when called.
+        initial_rate: Initial dropout rate to apply.
+
+    Usage:
+
+        base_layer = tf.keras.layers.Dense(units=3)
+        dropout_layer = ConcreteDropout(base_layer)
+        dropout_layer(inputs)
+
+    References:
+        Concrete Dropout. Gal et al. 2017. https://arxiv.org/abs/1705.07832
+    """
+
+    def __init__(
+        self,
+        layer,
+        initial_rate=0.1,
+        kernel_regularization=1e-6,
+        dropout_regularization=1e-5,
+        noise_shape=None,
+        seed=None,
+        **kwargs
+    ):
+        super(ConcreteDropout, self).__init__(layer, **kwargs)
+        self.initial_rate = initial_rate
+        self.kernel_regularization = kernel_regularization
+        self.dropout_regularization = dropout_regularization
+        self.noise_shape = noise_shape
+        self.seed = seed
+        self.supports_masking = True
+        self.rate_logit = None
+
+    def build(self, input_shape):
+        super(ConcreteDropout, self).build(input_shape)
+
+        initial_rate_norm = np.log(self.initial_rate) - np.log(1.0 - self.initial_rate)
```

Comment:
Can this be implemented with tensorflow to support tensors and numpy arrays?

Reply:
For example, I see a case where someone supplies a constant tensor to the initial_rate argument and this blows up. Or maybe we should assert something in the constructor OR put it in the docs and have a test for it, if we want to keep this.

Reply:
Definitely can use tf. I should probably add an op for the sigmoid inverse too. That’s generally useful.